src/couchdb/rolling-database.ts

```diff
@@ -33,109 +36,97 @@ export interface RollingFindOptions {
+      out.push({ id: _id, rev: _rev, db: name, doc: rest })
+    }
+
+    // Stop if we have enough results:
+    if (out.length >= limit) break
+
+    // Stop once we are in the past:
+    if (startDate != null && start.valueOf() < startDate.valueOf()) break
+  }
+  return out
 }
 
-export async function makeRollingDatabase(
+/**
+ * Inserts a document into the newest database in a rolling list of databases.
+ */
+export async function rollingInsert(
   connection: ServerScope,
-  databaseList: SyncedDocument<RollingDatabasesDocument>
-): Promise<RollingDatabase> {
+  databaseList: SyncedDocument<RollingDatabasesDocument>,
+  doc: CouchDoc<object> & { db?: string }
+): Promise<DocumentInsertResponse> {
+  // Simply insert if we know where it goes:
+  if (doc.db != null) {
+    return await connection.use(doc.db).insert({
+      ...doc.doc,
+      _id: doc.id,
+      _rev: doc.rev
+    })
+  }
+
+  // Otherwise, search for a database to put it in:
+  for (const { name, start } of databaseList.doc.databases) {
+    if (start.valueOf() > Date.now()) continue
+
+    return await connection.use(name).insert({
+      ...doc.doc,
+      _id: doc.id,
+      _rev: doc.rev
+    })
+  }
+
+  throw new Error('No rolling databases exist')
+}
+
+/**
+ * Sets up a collection of rolling databases.
+ */
+export async function setupRollingDatabases(
+  connection: ServerScope,
+  setupInfo: RollingDatabaseSetup,
+  opts: SetupDatabaseOptions = {}
+): Promise<void> {
+  const { databaseList } = setupInfo
+  const { log = console.log } = opts
+
   // We only want to set up each database once:
   const ready: { [name: string]: true } = {}
-
-  return {
-    async find(selector, opts = {}) {
-      const { limit = 20, startDate, sort } = opts
-
-      const out: Array<CouchDoc<object> & { db: string }> = []
-      for (const { name, start } of databaseList.doc.databases) {
-        if (!ready[name]) continue
-
-        const { docs } = await connection.use(name).find({
-          limit: limit - out.length,
-          selector,
-          sort
-        })
-        for (const doc of docs) {
-          const { _id, _rev, ...rest } = doc
-          out.push({ id: _id, rev: _rev, db: name, doc: rest })
-        }
-
-        // Stop if we have enough results:
-        if (out.length >= limit) break
-
-        // Stop once we are in the past:
-        if (startDate != null && start.valueOf() < startDate.valueOf()) break
-      }
-      return out
-    },
-
-    async insert(
-      doc: CouchDoc<object> & { db?: string }
-    ): Promise<DocumentInsertResponse> {
-      // Simply insert if we know where it goes:
-      if (doc.db != null) {
-        return await connection.use(doc.db).insert({
-          ...doc.doc,
-          _id: doc.id,
-          _rev: doc.rev
-        })
-      }
-
-      // Otherwise, search for a database to put it in:
-      for (const { name, start } of databaseList.doc.databases) {
-        if (!ready[name]) continue
```

Comment:
Didn't this provide some protection against race conditions?

Reply:
Not really. While it helps on a single core (but not completely), it's useless if you are running on multiple CPU cores or even multiple servers. Therefore, the only solution I could come up with is to always keep a "spare" database around with a start date in the future. That gives the nodes / cores / clusters plenty of time to prepare before switching over to the new database.

Since that's the only reliable stragegy, the `ready` check is just going to mask bugs, at best.